File: server/ai/minimax/minimax.go
package minimax

import (
	"chinese-chess/server/ai/minimax/eval"
	"chinese-chess/server/chess"
	"chinese-chess/server/shared"
	"chinese-chess/server/worker"
	"fmt"
	"math"
	"time"
)

type BestMove struct {
	move  chess.Move
	value int
}

type BestMoveWorker struct {
	board         *chess.Board
	moves         []chess.Move
	resultChannel chan BestMove
	search        func(board *chess.Board) int
}

func (bmw *BestMoveWorker) Task() {
	len := len(bmw.moves)

	fmt.Printf("BestMoveWorker Task (%d) start: %d\n", len, time.Now().Unix())

	bestMove := BestMove{
		move:  chess.Move{},
		value: -shared.INFINITE,
	}

	for _, move := range bmw.moves {
		x := move.From[0]
		y := move.From[1]
		piece := bmw.board.Cells[x][y]
		bmw.board.UpdatePiece(piece, move.To)
		value := bmw.search(bmw.board)
		bmw.board.BackMoves(1)
		if value > bestMove.value {
			bestMove.value = value
			bestMove.move.From = piece.Pos
			bestMove.move.To = move.To
		}
	}
	fmt.Printf("BestMoveWorker Task (%d) end: %d\n", len, time.Now().Unix())

	bmw.resultChannel <- bestMove
}

type MiniMax struct {
	depth     int
	evalModel eval.IEvalModel
	cutOff    bool
}

const WORKER_NUM = 1600

func getColor(isMax bool, color chess.Color) chess.Color {
	if isMax {
		return color
	} else {
		if color == chess.Red {
			return chess.Black
		} else {
			return chess.Red
		}
	}
}

func (miniMax *MiniMax) search(board *chess.Board, color chess.Color, depth int, isMax bool, alpha int, beta int) int {
	if depth == 0 || board.IsFinish() {
		return miniMax.evalModel.Eval(board, color)
	}

	var value int
	if isMax {
		value = -shared.INFINITE
	} else {
		value = shared.INFINITE
	}

	moves := board.GenerateMoves(getColor(isMax, color))
	for _, move := range moves {
		x := move.From[0]
		y := move.From[1]
		piece := board.Cells[x][y]
		board.UpdatePiece(piece, move.To)
		_value := miniMax.search(board, color, depth-1, !isMax, alpha, beta)
		board.BackMoves(1)
		if isMax {
			value = shared.Max(value, _value)
			if miniMax.cutOff {
				alpha = shared.Max(alpha, value)
				if alpha >= beta {
					return alpha
				}
			}
		} else {
			value = shared.Min(value, _value)
			if miniMax.cutOff {
				beta = shared.Min(beta, value)
				if alpha >= beta {
					return beta
				}
			}
		}
	}
	return value
}

func (miniMax *MiniMax) GetBestMove(board *chess.Board, color chess.Color, moves []chess.Move) BestMove {
	// // 单线程
	// return miniMax.search(board, color, miniMax.depth-1, false, -shared.INFINITE, shared.INFINITE)

	// 多线程
	len := len(moves)
	workerNum := WORKER_NUM
	if workerNum > len {
		workerNum = len
	}
	n := int(math.Floor(float64(len) / float64(workerNum)))
	remainder := len - workerNum*n

	p := worker.New(workerNum)
	maxChannel := make(chan BestMove)
	lastEnd := 0
	for i := 0; i < workerNum; i++ {
		start := lastEnd
		if start > len-1 {
			break
		}
		offset := n
		if remainder > 0 {
			offset += 1
		}
		remainder -= 1
		subArr := moves[start : start+offset]
		lastEnd = start + offset
		p.Run(&BestMoveWorker{board: board.Clone(), moves: subArr, search: func(board *chess.Board) int {
			return miniMax.search(board, color, miniMax.depth-1, false, -shared.INFINITE, shared.INFINITE)
		}, resultChannel: maxChannel})
	}

	bestMove := BestMove{
		move:  chess.Move{},
		value: -shared.INFINITE,
	}
	for i := 0; i < workerNum; i++ {
		subBestMove := <-maxChannel
		if subBestMove.value > bestMove.value {
			bestMove = subBestMove
		}
	}
	p.Shutdown()

	return bestMove
}

func (miniMax *MiniMax) GetNextMove(board *chess.Board, color chess.Color) chess.Move {
	fmt.Printf("GetNextMove start: %d\n", time.Now().Unix())
	piecesMoves := board.GenerateMoves(color)
	bestMove := miniMax.GetBestMove(board, color, piecesMoves)
	fmt.Printf("GetNextMove end: %d\n", time.Now().Unix())
	return bestMove.move
}

func NewMiniMax(depth int, evalModel eval.IEvalModel, cutOff bool) *MiniMax {
	return &MiniMax{depth, evalModel, cutOff}
}

File: wasm-chinese-chess/src/minimax.rs
pub mod weight;

use std::{
    cmp::{max, min},
    sync::{mpsc, Arc},
    thread,
};

use crate::{
    board::{Board, Move},
    shared::{Color, Pos},
};

use self::weight::WeightEvalModel;

struct BestMove {
    m: Move,
    value: i32,
}

fn get_color(is_max: bool, color: &Color) -> Color {
    if is_max {
        return *color;
    } else {
        if *color == Color::Red {
            return Color::Black;
        } else {
            return Color::Red;
        }
    }
}

fn search(
    cut_off: bool,
    board: &mut Board,
    color: &Color,
    depth: i32,
    is_max: bool,
    mut alpha: i32,
    mut beta: i32,
) -> i32 {
    if depth == 0 || board.is_finish() {
        return WeightEvalModel::eval(board, color);
    }

    let mut value: i32;
    if is_max {
        value = -INFINITE;
    } else {
        value = INFINITE;
    }

    let moves = board.generate_moves(&get_color(is_max, color));

    for m in moves.iter() {
        board.update_piece(&m.from, &m.to);
        // println!("move: {:?}, {:?}", m.from, m.to);
        // println!("{:}", board);
        let _value = search(cut_off, board, color, depth - 1, !is_max, alpha, beta);
        board.back_moves(1);
        // println!("back_move:");
        // println!("{:}", board);
        if is_max {
            value = max(value, _value);
            if cut_off {
                alpha = max(alpha, value);
                if alpha >= beta {
                    return alpha;
                }
            }
        } else {
            value = min(value, _value);
            if cut_off {
                beta = min(beta, value);
                if alpha >= beta {
                    return beta;
                }
            }
        }
    }
    return value;
}

pub struct MiniMax {
    depth: i32,
    cut_off: bool,
}

const INFINITE: i32 = 999999999;

impl MiniMax {
    pub fn search(
        &self,
        board: &mut Board,
        color: &Color,
        depth: i32,
        is_max: bool,
        mut alpha: i32,
        mut beta: i32,
    ) -> i32 {
        if depth == 0 || board.is_finish() {
            return WeightEvalModel::eval(board, color);
        }

        let mut value: i32;
        if is_max {
            value = -INFINITE;
        } else {
            value = INFINITE;
        }

        let moves = board.generate_moves(&get_color(is_max, color));

        for m in moves.iter() {
            board.update_piece(&m.from, &m.to);
            // println!("move: {:?}, {:?}", m.from, m.to);
            // println!("{:}", board);
            let _value = self.search(board, color, depth - 1, !is_max, alpha, beta);
            board.back_moves(1);
            // println!("back_move:");
            // println!("{:}", board);
            if is_max {
                value = max(value, _value);
                if self.cut_off {
                    alpha = max(alpha, value);
                    if alpha >= beta {
                        return alpha;
                    }
                }
            } else {
                value = min(value, _value);
                if self.cut_off {
                    beta = min(beta, value);
                    if alpha >= beta {
                        return beta;
                    }
                }
            }
        }
        return value;
    }

    pub fn get_best_move(&self, board: &mut Board, color: &Color, moves: Vec<Move>) -> BestMove {
        let mut best_move = BestMove {
            value: -INFINITE,
            m: Move {
                from: Pos(0, 0),
                to: Pos(0, 0),
            },
        };
        println!("{:?}", moves.len());
        let thread_num = 8;
        let chunks = moves.len().div_ceil(thread_num);
        let moves_ref = Arc::new(moves);

        let (tx, rx) = mpsc::channel::<(Move, i32)>();
        let cut_off = self.cut_off;
        let depth = self.depth;
        for i in 0..moves_ref.chunks(chunks).count() {
            let moves_clone = moves_ref.clone();
            let mut board_clone = board.clone();
            let color_clone = color.clone();
            let tx_clone = tx.clone();
            thread::spawn(move || {
                for moves in moves_clone.chunks(thread_num).nth(i) {
                    for m in moves.iter() {
                        board_clone.update_piece(&m.from, &m.to);
                        let value = search(
                            cut_off,
                            &mut board_clone,
                            &color_clone,
                            depth - 1,
                            false,
                            -INFINITE,
                            INFINITE,
                        );

                        board_clone.back_moves(1);

                        tx_clone.send((m.clone(), value)).unwrap();
                    }
                }
            });
        }

        for _ in 0..thread_num {
            match rx.recv() {
                Ok((m, value)) => {
                    println!("{:?}, {:?}", m, value);
                    if best_move.value < value {
                        best_move.m = m;
                        best_move.value = value;
                    }
                }
                Err(_) => todo!(),
            }
        }

        return best_move;
    }

    pub fn get_next_move(&self, board: &mut Board, color: &Color) -> Move {
        let pieces_moves = board.generate_moves(color);
        let best_move = self.get_best_move(board, color, pieces_moves);
        println!("{}", best_move.value);
        best_move.m
    }

    pub fn new(depth: i32, cut_off: bool) -> MiniMax {
        MiniMax { depth, cut_off }
    }
}
